### PR TITLE
Implement support for Github OAuth 2.0 provider in UAA

### PR DESCRIPTION
**Pull request by bgandon**
Hi Cloud Foundry fellows,

I've implemented support for Github OAuth 2 as Identity Provider (IdP). This fixes the issue described in #798 .

All 4600+ tests are passing. (I didn't fork from the latests `develop` HEAD because tests are currently not passing there.)
I've added some unit tests to cover the main Github use-case OAuth 2.0 “Autorization Code Grant” flow.
There is also some documentation showing how to use it.
I've deployed this version of UAA in a Cloud Foundry, making a fork of the `uaa-release`. Testing the result my own Gstack's sandbox CF environment, the real-life Github authentication works fine.

As a design choice that could be discussed, I've re-used the `checkTokenUrl` property (that is not used anywhere), in order to implement the “user-info” endpoint.
In #798, the `userInfoUrl` config is suggested and definitely looks more appropriate. I'll submit some improvement soon to address this. (**edit:** this is pushed now, see below)

Best,
Benjamin

### CODE REVIEW

**Comment by bgandon** [comment]
Hi folks and dear UAA maintainers,
I've pushed a refactoring of the `userInfoUrl` property so that it can apply to both OIDC and OAuth2 Identity Providers.
With this change, I could start using this `userInfoUrl` property when configuring Github OAuth2 IDP and leverage it to fetch  user metadata.
All unit tests are passing locally with `./gradlew build`.

**Comment by bgandon** [comment]
Hi Markus @strehle,
I've rebased my work on top of the latest `develop` branch, with some minor updates for it to work.
Gradle build and tests are properly passing.
Now I need to fix things with the CLA and my company membership to the CFF…
Best,
Benjamin

**Comment by bgandon** [comment]
Did you find time to look at it? (while I'm fixing the CLA thing… still pending sorry)

File: model/src/main/java/org/cloudfoundry/identity/uaa/provider/AbstractExternalOAuthIdentityProviderDefinition.java
Comment (strehle) [inline model/src/main/java/org/cloudfoundry/identity/uaa/provider/AbstractExternalOAuthIdentityProviderDefinition.java:31]: This parameter should be added to documentation, see
https://docs.cloudfoundry.org/api/uaa/version/75.2.0/index.html#oauth-oidc

**Review by strehle — COMMENTED** [review]
Please finish CLA, then we can finish it

**Comment by bgandon** [comment]
I've just signed the CLA at corporate level. Wondering how to refresh here so that it is properly taken into account… maybe some commits? … or closing & re-opening the PR?

File: model/src/main/java/org/cloudfoundry/identity/uaa/provider/AbstractExternalOAuthIdentityProviderDefinition.java
Comment (bgandon) [reply model/src/main/java/org/cloudfoundry/identity/uaa/provider/AbstractExternalOAuthIdentityProviderDefinition.java:31]: Sure! I had only given the canonical config for Github in `docs/github-oauth2-provider.md`.

All right, I've located the source of the slate API docs and pushed some documentation for this property!

**Comment by torsten-sap** [comment]
@bgandon If you click on the EasyCLA logo and follow the instructions it should turn green.

At least in my case reopening was not required a few month ago.

**Comment by strehle** [comment]
@bgandon  you can try a dummy PR if you think CLA is failing only here, but please understand we need to have green agreement here. if you have corporate CLA maybe your org. CLA manager needs to approve

**Comment by bgandon** [comment]
Thanks @torsten-sap, I finally managed to get it work and it turned green.

**Review by strehle — APPROVED** [review]

**Comment by bgandon** [comment]
Thank you @strehle for approving!
What's the next step? Do we need one more approval before merging?

**Comment by strehle** [comment]
> What's the next step? Do we need one more approval before merging?

I tested manually your github-oauth2-provider.md

From YAML the settings are not filled correctly becuase missing

server/src/main/java/org/cloudfoundry/identity/uaa/provider/oauth/OauthIDPWrapperFactoryBean.java
Line 128
add

idpDefinition.setUserInfoUrl(idpDefinitionMap.get("userInfoUrl") == null ? null : new URL((String) idpDefinitionMap.get("userInfoUrl")));

Because then from YAML the json config is filled into DB correctly and later the flow can be used.

I assume you tested before only creating an IDP from REST.

**Comment by bgandon** [comment]
I've used the provided `userInfoUrl: https://api.github.com/user` settings, which is mandatory for properly fetching user details with Github.
See: https://github.com/cloudfoundry/uaa/pull/1463/files#diff-111e27ec77818be646f5a9eaa9e8612809f58606834180c9c3f3cb862d02ef77R27

Oh, I remember now!
I've patched the UAA BOSH Release.
But when I look at my patched branch, I only see this line added at the beginning of  `jobs/uaa/templates/config/uaa.yml.erb`:
```
require 'uri'
```
See: https://github.com/gstackio/gk-uaa-boshrelease/blob/feature/github-oauth2/jobs/uaa/templates/config/uaa.yml.erb#L2

If I remember correctly, this was a pre-existing bug that prevented proper setup of UAA…
Hope this may help you test manually.
At least with my patched release I can certify it works. Not very fine, because Github username cannot be properly split into firstname/givenname in the UAA, but I can tell people from Github can login. With proper SSO setup for Stratos, it's _wunderbar_! 😉

**Comment by strehle** [comment]
> patched the UAA

ok, but I assume you want that it works also outside of bosh , e.g. k8s or any other standalone scenario ?

I run uaa.war simply in a docker compose env but also start it via ./gradlew run . With this simply the uaa.yml / login.yml is read and there is see that setting userInfoUrl is not taken because of missing

`idpDefinition.setUserInfoUrl(idpDefinitionMap.get("userInfoUrl") == null ? null : new URL((String) idpDefinitionMap.get("userInfoUrl")));`

 setter after
https://github.com/cloudfoundry/uaa/blob/develop/server/src/main/java/org/cloudfoundry/identity/uaa/provider/oauth/OauthIDPWrapperFactoryBean.java#L125-L127

So if you dont provide any further commit than ok for me/us , then we can merge

**Comment by bgandon** [comment]
Oh, sorry I haven't had time to answer your message. I think you've seen something I've missed.

Indeed, I've only tested with the patched BOSH Release, and it also has to run properly in Kubernetes and standalone scenarios.

Now that this is merged, shall I submit a new PR in order to add the missing line of code that you mentioned?

**Comment by strehle** [comment]
see 
https://github.com/cloudfoundry/uaa/pull/1595/files
please test your self with
1. add https://github.com/cloudfoundry/uaa/blob/develop/docs/github-oauth2-provider.md in login.yml and store it in folder https://github.com/cloudfoundry/uaa/tree/develop/scripts/cargo
2. start local uaa with ./gradlew run
3. open http://localhost:8080/uaa

**Comment by bgandon** [comment]
Ok great! Thank you!

**Pull request merged**
